<a href="https://colab.research.google.com/github/alexmjn/DS-Unit-2-Linear-Models/blob/master/module3-ridge-regression/LS_DS_213_assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Lambda School Data Science

*Unit 2, Sprint 1, Module 3*

---

# Ridge Regression

## Assignment

We're going back to our other **New York City** real estate dataset. Instead of predicting apartment rents, you'll predict property sales prices.

But not just for condos in Tribeca...

- [ ] Use a subset of the data where `BUILDING_CLASS_CATEGORY` == `'01 ONE FAMILY DWELLINGS'` and the sale price was more than 100 thousand and less than 2 million.
- [ ] Do train/test split. Use data from January — March 2019 to train. Use data from April 2019 to test.
- [ ] Do one-hot encoding of categorical features.
- [ ] Do feature selection with `SelectKBest`.
- [ ] Fit a ridge regression model with multiple features. Use the `normalize=True` parameter (or do [feature scaling](https://scikit-learn.org/stable/modules/preprocessing.html) beforehand — use the scaler's `fit_transform` method with the train set, and the scaler's `transform` method with the test set)
- [ ] Get mean absolute error for the test set.
- [ ] As always, commit your notebook to your fork of the GitHub repo.

The [NYC Department of Finance](https://www1.nyc.gov/site/finance/taxes/property-rolling-sales-data.page) has a glossary of property sales terms and NYC Building Class Code Descriptions. The data comes from the [NYC OpenData](https://data.cityofnewyork.us/browse?q=NYC%20calendar%20sales) portal.


## Stretch Goals

Don't worry, you aren't expected to do all these stretch goals! These are just ideas to consider and choose from.

- [ ] Add your own stretch goal(s) !
- [ ] Instead of `Ridge`, try `LinearRegression`. Depending on how many features you select, your errors will probably blow up! 💥
- [ ] Instead of `Ridge`, try [`RidgeCV`](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.RidgeCV.html).
- [ ] Learn more about feature selection:
    - ["Permutation importance"](https://www.kaggle.com/dansbecker/permutation-importance)
    - [scikit-learn's User Guide for Feature Selection](https://scikit-learn.org/stable/modules/feature_selection.html)
    - [mlxtend](http://rasbt.github.io/mlxtend/) library
    - scikit-learn-contrib libraries: [boruta_py](https://github.com/scikit-learn-contrib/boruta_py) & [stability-selection](https://github.com/scikit-learn-contrib/stability-selection)
    - [_Feature Engineering and Selection_](http://www.feat.engineering/) by Kuhn & Johnson.
- [ ] Try [statsmodels](https://www.statsmodels.org/stable/index.html) if you’re interested in more inferential statistical approach to linear regression and feature selection, looking at p values and 95% confidence intervals for the coefficients.
- [ ] Read [_An Introduction to Statistical Learning_](http://faculty.marshall.usc.edu/gareth-james/ISL/ISLR%20Seventh%20Printing.pdf), Chapters 1-3, for more math & theory, but in an accessible, readable way.
- [ ] Try [scikit-learn pipelines](https://scikit-learn.org/stable/modules/compose.html).

In [0]:
%%capture
import sys

# If you're on Colab:
if 'google.colab' in sys.modules:
    DATA_PATH = 'https://raw.githubusercontent.com/LambdaSchool/DS-Unit-2-Applied-Modeling/master/data/'
    !pip install category_encoders==2.*

# If you're working locally:
else:
    DATA_PATH = '../data/'
    
# Ignore this Numpy warning when using Plotly Express:
# FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
import warnings
warnings.filterwarnings(action='ignore', category=FutureWarning, module='numpy')

In [0]:
import pandas as pd
import pandas_profiling

# Read New York City property sales data
df = pd.read_csv(DATA_PATH+'condos/NYC_Citywide_Rolling_Calendar_Sales.csv')

# Change column names: replace spaces with underscores
df.columns = [col.replace(' ', '_') for col in df]

# SALE_PRICE was read as strings.
# Remove symbols, convert to integer
df['SALE_PRICE'] = (
    df['SALE_PRICE']
    .str.replace('$','')
    .str.replace('-','')
    .str.replace(',','')
    .astype(int)
)

In [0]:
# BOROUGH is a numeric column, but arguably should be a categorical feature,
# so convert it from a number to a string
df['BOROUGH'] = df['BOROUGH'].astype(str)

In [0]:
# Reduce cardinality for NEIGHBORHOOD feature

# Get a list of the top 10 neighborhoods
top10 = df['NEIGHBORHOOD'].value_counts()[:10].index

# At locations where the neighborhood is NOT in the top 10, 
# replace the neighborhood with 'OTHER'
df.loc[~df['NEIGHBORHOOD'].isin(top10), 'NEIGHBORHOOD'] = 'OTHER'

In [5]:
df.head().T

0  ...                                  4
BOROUGH                                                       1  ...                                  1
NEIGHBORHOOD                                              OTHER  ...            UPPER EAST SIDE (59-79)
BUILDING_CLASS_CATEGORY         13 CONDOS - ELEVATOR APARTMENTS  ...  15 CONDOS - 2-10 UNIT RESIDENTIAL
TAX_CLASS_AT_PRESENT                                          2  ...                                 2C
BLOCK                                                       716  ...                               1379
LOT                                                        1246  ...                               1402
EASE-MENT                                                   NaN  ...                                NaN
BUILDING_CLASS_AT_PRESENT                                    R4  ...                                 R1
ADDRESS                             447 WEST 18TH STREET, PH12A  ...             20 EAST 65TH STREET, B
APARTMENT_NUMBER                                          PH12A  ...                                  B
ZIP_CODE                                                  10011  ...                              10065
RESIDENTIAL_UNITS                                             1  ...                                  1
COMMERCIAL_UNITS                                              0  ...                                  0
TOTAL_UNITS                                                   1  ...                                  1
LAND_SQUARE_FEET                                         10,733  ...                                  0
GROSS_SQUARE_FEET                                          1979  ...                               6406
YEAR_BUILT                                                 2007  ...                                  0
TAX_CLASS_AT_TIME_OF_SALE                                     2  ...                                  2
BUILDING_CLASS_AT_TIME_OF_SALE                               R4  ...                                 R1
SALE_PRICE                                                    0  ...                                  0
SALE_DATE                                            01/01/2019  ...                         01/01/2019

[21 rows x 5 columns]

In [0]:
df = df[df["BUILDING_CLASS_CATEGORY"] == '01 ONE FAMILY DWELLINGS']
df = df[df["SALE_PRICE"].between(100000, 2000000)]
df = df.drop(["EASE-MENT", "ADDRESS", "APARTMENT_NUMBER"], axis = 1)

In [7]:
df.dtypes

BOROUGH                            object
NEIGHBORHOOD                       object
BUILDING_CLASS_CATEGORY            object
TAX_CLASS_AT_PRESENT               object
BLOCK                               int64
LOT                                 int64
BUILDING_CLASS_AT_PRESENT          object
ZIP_CODE                          float64
RESIDENTIAL_UNITS                 float64
COMMERCIAL_UNITS                  float64
TOTAL_UNITS                       float64
LAND_SQUARE_FEET                   object
GROSS_SQUARE_FEET                 float64
YEAR_BUILT                        float64
TAX_CLASS_AT_TIME_OF_SALE           int64
BUILDING_CLASS_AT_TIME_OF_SALE     object
SALE_PRICE                          int64
SALE_DATE                          object
dtype: object

Looks like we need to cast LAND_SQUARE_FEET as an integer.

In [0]:
df["LAND_SQUARE_FEET"] = df["LAND_SQUARE_FEET"].str.replace(",","").astype(int)

In [0]:
categoricals = ["BOROUGH", "NEIGHBORHOOD", "TAX_CLASS_AT_PRESENT", "BUILDING_CLASS_AT_PRESENT", "ZIP_CODE", "BUILDING_CLASS_AT_TIME_OF_SALE"]

In [0]:
df["SALE_DATE"] = pd.to_datetime(df["SALE_DATE"], infer_datetime_format=True)

In [0]:
cutoff = pd.to_datetime("2019-04-01")

In [0]:
train = df[df["SALE_DATE"] < cutoff]
test = df[df["SALE_DATE"] >= cutoff]

In [13]:
df.columns[df.nunique() <= 1]

Index(['BUILDING_CLASS_CATEGORY', 'TAX_CLASS_AT_TIME_OF_SALE'], dtype='object')

Drop all columns which have only one unique value, as they don't provide any extra information.

In [0]:
features = df.columns.drop(["SALE_PRICE", "SALE_DATE", "BUILDING_CLASS_CATEGORY", "TAX_CLASS_AT_TIME_OF_SALE"])
target = "SALE_PRICE"

In [0]:
X_train = train[features]
X_test = test[features]
y_train = train[target]
y_test = test[target]

In [0]:
import category_encoders as ce

In [0]:
encoder = ce.OneHotEncoder(use_cat_names = True, cols = categoricals)

In [0]:
X_train = encoder.fit_transform(X_train)

In [19]:
X_train.shape

(2517, 168)

In [0]:
X_test = encoder.transform(X_test)

In [0]:
from sklearn.feature_selection import SelectKBest, f_regression
from sklearn.metrics import mean_absolute_error
from sklearn.linear_model import LinearRegression

In [22]:
for k in range(1, len(X_train.columns) + 1):
  print(f'{k} features')

  selector = SelectKBest(score_func = f_regression, k = k)
  X_train_selected = selector.fit_transform(X_train, y_train)
  X_test_selected = selector.transform(X_test)

  model = LinearRegression()
  model.fit(X_train_selected, y_train)
  y_pred = model.predict(X_test_selected)
  mae = mean_absolute_error(y_test, y_pred)

  print(f'Test MAE: {mae:.0f} \n')

1 features
Test MAE: 185788 

2 features
Test MAE: 181871 

3 features
Test MAE: 181603 

4 features
Test MAE: 181603 

5 features
Test MAE: 173029 

6 features
Test MAE: 172716 

7 features
Test MAE: 171356 

8 features
Test MAE: 171025 

9 features
Test MAE: 169695 

10 features
Test MAE: 169660 

11 features
Test MAE: 166936 

12 features
Test MAE: 155140 

13 features
Test MAE: 154675 

14 features
Test MAE: 155261 

15 features
Test MAE: 154705 

16 features
Test MAE: 153055 

17 features
Test MAE: 151429 

18 features
Test MAE: 153482 

19 features
Test MAE: 153484 

20 features
Test MAE: 153486 

21 features
Test MAE: 154867 

22 features
Test MAE: 154755 

23 features
Test MAE: 153871 

24 features
Test MAE: 151709 

25 features
Test MAE: 151435 

26 features
Test MAE: 151254 

27 features
Test MAE: 151254 

28 features
Test MAE: 151268 

29 features
Test MAE: 150075 

30 features
Test MAE: 149181 

31 features
Test MAE: 149181 

32 features
Test MAE: 150487 

33 features
Test 

Looks like the optimal model has 98 features. 

In [0]:
selector = SelectKBest(score_func = f_regression, k = 98)
X_train_selected = selector.fit_transform(X_train, y_train)
X_test_selected = selector.transform(X_test)

## Ridge Regression

In [0]:
from sklearn.linear_model import Ridge, RidgeCV
from sklearn.preprocessing import StandardScaler
#scaler = StandardScaler()
#X_train_selected = scaler.fit_transform(X_train_selected)
#X_test_selected = scaler.transform(X_test_selected)

In [25]:
X_train_selected

array([[1., 0., 0., ..., 0., 0., 0.],
       [0., 1., 0., ..., 0., 0., 0.],
       [0., 0., 1., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 1., 0., ..., 0., 0., 0.]])

In [0]:
alphas = [.01, .1, .5, 1, 5, 10]

In [27]:
ridge_model = RidgeCV(alphas = alphas, normalize = True, store_cv_values=True)
ridge_model.fit(X_train_selected, y_train)

RidgeCV(alphas=array([ 0.01,  0.1 ,  0.5 ,  1.  ,  5.  , 10.  ]), cv=None,
        fit_intercept=True, gcv_mode=None, normalize=True, scoring=None,
        store_cv_values=True)

In [0]:
y_pred = ridge_model.predict(X_test_selected)
mae = mean_absolute_error(y_test, y_pred)

In [29]:
mae

139380.1508588058

This seems to be the same MAE as without scaling? 138286

Nope: MAE with scaling 139380.

In [30]:
ridge_model.cv_values_

array([[2.08132878e+11, 1.80753055e+11, 1.18645463e+11, 8.35922172e+10,
        2.52414195e+10, 1.48672430e+10],
       [2.86840687e+11, 2.75209636e+11, 2.49483511e+11, 2.33927328e+11,
        2.00702646e+11, 1.91316496e+11],
       [4.23406994e+10, 3.92980930e+10, 3.61758328e+10, 3.55616184e+10,
        3.51590561e+10, 3.52333047e+10],
       ...,
       [1.67224808e+10, 1.45849810e+10, 8.18975614e+09, 4.41507018e+09,
        1.99853381e+08, 5.67288359e+06],
       [6.37380130e+08, 2.01925537e+08, 2.08806486e+08, 1.11848148e+09,
        5.58712395e+09, 7.37747071e+09],
       [1.45316032e+11, 1.51198562e+11, 1.73189442e+11, 1.91563182e+11,
        2.37652851e+11, 2.50486550e+11]])

In [31]:
ridge_model.alpha_

0.1